### PR TITLE
add tarred datasets

### PR DESCRIPTION
**Pull request by stevehuang52**
# What does this PR do ?

Add tarred dataset, bucketing datasets are already supported as well.

### CODE REVIEW

File: nemo/collections/multimodal/models/speechllm_models.py [hunk 1]
```diff
@@ -433,10 +451,75 @@ def _build_dataset(self, data_cfg, is_train=True):
...
+        logging.info(f'Building dataloader with consumed samples: {consumed_samples}')
+        if isinstance(dataset, BlendableDataset):
+            collate_fn = dataset.datasets[0].collate_fn
+        elif hasattr(dataset, 'collate_fn'):
+            collate_fn = dataset.collate_fn
+        elif hasattr(dataset.datasets[0], 'collate_fn'):
+            # support datasets that are lists of entries
+            collate_fn = dataset.datasets[0].collate_fn
+        else:
+            # support datasets that are lists of lists
```
Comment (zhehuaichen) [inline nemo/collections/multimodal/models/speechllm_models.py:485]: could you explain why we need these for supporting tar?

File: nemo/collections/multimodal/data/audio_text_qa_dataset.py [hunk 2]
```diff
@@ -375,78 +334,208 @@ def _process_example(self, context: str, output: str):
...
+        truncation_field: str = "answer",
+        pad_to_max_length: bool = False,  # (@adithyare) allows for much faster training especially in PEFT settings.
+        prompt_template: str = None,
+        virtual_tokens: int = 0,
+        tokens_to_generate: int = 0,
+        index_by_file_id: bool = False,
+        input_key: str = 'input',
+        output_key: str = 'output',
+        end_string: Optional[str] = None,
+        question_file: Optional[Union[str, List[str]]] = None,
```
Comment (zhehuaichen) [inline nemo/collections/multimodal/data/audio_text_qa_dataset.py:408]: question for myself and no need to be in this PR: could you provide an example of using question_file

File: workspace/run_sft_tarred.sh [hunk 3]
```diff
@@ -0,0 +1,46 @@
+NEMO_DIR=/home/heh/codes/nemo-zhehuai
```
Comment (zhehuaichen) [inline workspace/run_sft_tarred.sh:1]: could you test locally to make sure both tar and untar cases work

**Review by zhehuaichen — CHANGES_REQUESTED** [review]
Thank you so much! Could you add to the description in this PR as TODOs on what needs to be done to support bucket dataset?

File: examples/multimodel/conf/speechllm/modularized_speech_gpt_config.yaml [hunk 4]
```diff
@@ -45,7 +45,7 @@ exp_manager:
     save_top_k: 1
     mode: min
     save_nemo_on_train_end: True
-    filename: '${name}--{${exp_manager.checkpoint_callback_params.monitor}:.3f}-{step}'
+    filename: '${name}--{${exp_manager.checkpoint_callback_params.monitor}:.3f}-{step}-{epoch}'
```
[the remark this answers is not in the document]
Comment (stevehuang52) [reply examples/multimodel/conf/speechllm/modularized_speech_gpt_config.yaml:48]: to add the epoch number in ckpt name

File: nemo/collections/multimodal/data/audio_text_qa_dataset.py (hunk 2 above)
Comment (stevehuang52) [reply nemo/collections/multimodal/data/audio_text_qa_dataset.py:408]: not implemented, just placeholder

File: nemo/collections/multimodal/models/speechllm_models.py (hunk 1 above)
Comment (stevehuang52) [reply nemo/collections/multimodal/models/speechllm_models.py:485]: followed ASR tarred datasets, I think it's for bucketing

File: examples/multimodel/conf/speechllm/modularized_speech_gpt_config.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (zhehuaichen) [reply examples/multimodel/conf/speechllm/modularized_speech_gpt_config.yaml:48]: do we need to load any other variable/info from the ckpt or that is just a naming different?

File: workspace/run_sft_tarred.sh (hunk 3 above)
Comment (stevehuang52) [reply workspace/run_sft_tarred.sh:1]: fixed sample_alpha and tested on both tar and untar datasets

**Review by zhehuaichen — APPROVED** [review]
Nice work! Thank you so much!

File: examples/multimodel/conf/speechllm/modularized_speech_gpt_config.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (stevehuang52) [reply examples/multimodel/conf/speechllm/modularized_speech_gpt_config.yaml:48]: just a naming issue

**Pull request merged**